src/cluster.c

```diff
@@ -5875,6 +6003,32 @@ NULL
         } else if (!strcasecmp(c->argv[3]->ptr,"node") && c->argc == 5) {
             /* CLUSTER SETSLOT <SLOT> NODE <NODE ID> */
             n = clusterLookupNode(c->argv[4]->ptr, sdslen(c->argv[4]->ptr));
+
+            /* When finalizing the slot, there is a possibility that the
+             * destination node B sends cluster PONG to the source node A
+             * before SETSLOT has been replicated to B'. If B crashes here,
+             * B' will be in importing state and the slot will have no owner.
+             * To help mitigate this issue, we enforce the following order
+             * for slot migration finalization such that the replicas will
+             * finalize the slot ownership before this primary:
```

Comment:
An alternative is we could have some threshold after which if a node is importing data, and no other node in the cluster owns that data, we just take ownership of the slot. This removes the need for complex state handling with the replica.

Reply:
> what if the cluster was partitioned around the same time and the original owner was out of touch temporarily?

This is only a problem for availability, since it should still know about the ownership in the other slots. If as far as the node knows, it doesn't own the slots so it won't take anything over.

> How does the importing node know definitely within a deterministic time window that no one else owns the slot?

We don't know anything definitely, we have no strong forms of consistency without cluster v2.

> Until the ownership is fixed, the slot in question is unavailable.

This is true.

> I believe it's justified since it solves the problem at its core and provides the necessary determinism.

I know I've stated this before, but I don't believe this is really the core problem. The core of the problem is we have no consistent store of cluster state, everyone has their own local opinion of the state. There are also plenty of other availability drawbacks here:
1. If a replica is not currently responsive this will fail, whereas it could have completed successfully before. This is a different type of availability. My understanding is that the redis-cli finalizes the slot on all nodes together.
2. We can still lose the state. We aren't durably storing state, so we can end up with no one serving the slot. We can also end up with the state finalized on one of multiple replicas. That replica could win the election, and then subsequently fail, and the replica without the finalized state taking over. 

> This is the tricky part to get right reliably.

I'm happy as long as we converge.

@ushachar I would like you to take a look at this section.

Reply:
>I know I've stated this before, but I don't believe this is really the core problem. The core of the problem is we have no consistent store of cluster state, everyone has their own local opinion of the state. 

Fully agreed that strong consistency for (cluster) metadata is much more desired. I have an implicit assumption that we will need to continue working with cluster V1 in the foreseeable future even after cluster V2 materializes hence my definition of "core" problems came with that constraint. These are real pain points for the cluster V1 customers.